Drop LIKELY() / UNLIKELY() macros

cdumez opened the pull request:
#### 495585188e97480fc20b4f75c7d605edfb2d7140
<pre>
Drop LIKELY() / UNLIKELY() macros
<a href="https://bugs.webkit.org/show_bug.cgi?id=292559">https://bugs.webkit.org/show_bug.cgi?id=292559</a>

Reviewed by Ryosuke Niwa.

Drop LIKELY() / UNLIKELY() macros now that our code base relies on
[[likely]] / [[unlikely]] instead.

* Source/WTF/wtf/Assertions.h:
* Source/WTF/wtf/Compiler.h:

Canonical link: <a href="https://commits.webkit.org/294535@main">https://commits.webkit.org/294535@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/682a8831b303a4f630d48cbf00b2eb26fd8a7ef4

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/102211 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/21879 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/12194 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/107370 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/52847 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/22187 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/30386 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/77781 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/52847 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/105217 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/17169 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/92269 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/58117 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/16999 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/10297 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/52205 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/94882 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/86834 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/10367 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/109746 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/100820 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/29343 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/21632 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/86760 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/29706 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/88472 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/86344 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/21962 "Built successfully and passed tests") | [  ~~🧪 vision-wk2~~](https://ews-build.webkit.org/#/builders/126/builds/31152 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/8871 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/23585 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/29271 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/34566 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/124446 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/29082 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/34556 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/32405 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/30641 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WTF/wtf/Assertions.h:
```diff
@@ -372,6 +372,11 @@ WTF_EXPORT_PRIVATE NO_RETURN_DUE_TO_CRASH void WTFCrashWithSecurityImplication(v
 
+/* This header may be used in C code so we cannot rely on [[unlikely]]. */
+#if !defined(C_UNLIKELY_IMPL)
+#define C_UNLIKELY_IMPL(x) __builtin_expect(!!(x), 0)
```
Why do we need to declare this macro separately??

cdumez replied:
Assertions.h and the ASSERT() macros rely on this macro and are used from C code, where we cannot use `[[likely]]` / `[[unlikely]]`. My plan is to get rid of the LIKELY() / UNLIKELY() macros in Compiler.h once I've have ported the couple of call sites in Safari (.m files) to something else. So then, only this `C_UNLIKELY_IMPL` would remain, with a bad name, so people don't adopt it.

One alternative would be to keep the `LIKELY() / UNLIKELY()` macros for C code long term in Compiler.h. However, then I'd need to update assertions in Assertions.h to generate different code in C and C++. In C, it would call the `LIKELY() / UNLIKELY()` macros and in C++, it would use [[likely]] / [[unlikely]]. This didn't seem worth the trouble to me.

rniwa replied:
Hm... ok. I'm not sure `_IMPL` is adding much value here though. Maybe just `C_UNLIKELY` would be ok. Or maybe we should signify the fact it should be only used in this header file, in which case, maybe we can prefix with assert like: `ASSERT_C_UNLIKELY`.

cdumez replied:
`ASSERT_C_UNLIKELY` is a bit confusing in terms of naming since it doesn't assert. Maybe `ASSERTIONS_C_UNLIKELY` ? Or `UNLIKELY_FOR_ASSERTIONS`?

cdumez replied:
I updated the PR (it uses `UNLIKELY_FOR_C_ASSERTIONS` in Assertions.h and `LIKELY()` is completely gone since unused). Let me know if you're happy with this latest proposal.

rniwa replied:
sure, that sounds good to me.

cdumez committed:
Drop LIKELY() / UNLIKELY() macros
https://bugs.webkit.org/show_bug.cgi?id=292559

Reviewed by Ryosuke Niwa.

Drop LIKELY() / UNLIKELY() macros now that our code base relies on
[[likely]] / [[unlikely]] instead.

* Source/WTF/wtf/Assertions.h:
* Source/WTF/wtf/Compiler.h:

Canonical link: https://commits.webkit.org/294535@main
